### ISSUE TITLE
关于实验细节的疑问

### ISSUE THREAD

**Issue by fourbeans** [SinoLeu/CKDSNN#3]
您好，关注到您的工作，这篇文章的思路新颖有趣。对于实验结果我有以下问题向您请教：

1. 因为您未公开教师ANN模型的精度。根据您的方法，在同数据集、同backbone的实验中，学生SNN模型是否有超越教师ANN的模型表现？
2. 是否可以提供预训练好的教师ANN的模型权重方便复现和引用您的工作？

**Comment by SinoLeu** [comment]
您好，非常感谢您对我们工作的关注，也感谢您仔细阅读论文和代码。针对您的几个问题，简单回复如下。

**1. 关于学生 SNN 是否能够超过教师 ANN**

这个问题并没有一个固定答案，主要取决于 teacher checkpoint 的性能、student capacity、时间步以及具体的训练配置。

知识蒸馏中 teacher 主要提供额外的 feature/logit supervision，而 student 同时仍然受到真实标签的监督，因此 teacher 的准确率并不是 student 的严格性能上界。在 teacher 本身并非最优、student 使用了较好的正则化或训练策略等情况下，student 是有可能超过 teacher 的；而当 teacher 已经是一个充分训练且性能较强的模型时，考虑到 SNN 的时间离散化、脉冲表示等限制，student 通常还是会略低于 teacher。

因此，我们更关注的是通过 CKDSNN 缩小 ANN 与 SNN 之间的性能差距，而不是要求 student 必须超过 teacher。

**2. 关于教师 ANN 的预训练权重**

教师模型本身没有使用特殊的私有预训练方式，主要采用公开的 pretrained ANN 权重。

对于 ResNet，可以直接使用 `timm`、torchvision 或 Hugging Face 上对应的 ImageNet pretrained model。

对于 ViT也是类似的。例如 ViT-S 可以直接采用 `timm` / Hugging Face 中公开的 ImageNet pretrained checkpoint，然后在 CIFAR-100 上进行 fine-tuning，得到 CIFAR-100 teacher，再用于后续的 SNN distillation。`timm` 中同一 ViT 架构通常还提供不同 pretrained recipe，因此复现时建议固定具体的 checkpoint/tag。

例如可以从：

```python
import timm

model = timm.create_model(
    'vit_small_patch16_224',
    pretrained=True,
    num_classes=100
)
```

开始，在 CIFAR-100 上完成 teacher fine-tuning 后再进行 CKDSNN 蒸馏。

如果希望直接参考已有的 CIFAR-100 ViT-S teacher，也有一些公开工作提供了对应 checkpoint，例如 RSD 的公开代码中提供了 `ViT-S` 的 CIFAR-100 pretrained teacher，可作为复现实验时的参考。

**3. 关于 ImageNet 大规模实验的 config**

ImageNet 的整体训练配置基本遵循已有 SNN 大规模训练工作的设置，其中很多训练和数据处理配置可以参考以下两个 repository：

* `brain-intelligence-lab/temporal_efficient_training`
* `fangwei123456/Spike-Element-Wise-ResNet`

其中我们部分 CIFAR/CIFAR-100 的训练方式、数据增强和优化设置 follow `temporal_efficient_training` 的实现；ImageNet 上的 SEW-ResNet 设置则主要参考 SEW-ResNet 官方的大规模训练 recipe。前者公开代码中包括 CIFAR-100 的 RandomCrop、RandomHorizontalFlip 和 normalization 等处理。

ImageNet 可以参考如下配置：

```yaml
dataset: imagenet
student: sew_resnet18        # or sew_resnet34
teacher: resnet18            # corresponding ANN teacher

T: 4
connect_f: ADD

optimizer: SGD
lr: 0.1
momentum: 0.9
weight_decay: 1e-4

scheduler: cosine
epochs: 320
amp: true
```

这部分基本 follow SEW-ResNet 的 ImageNet 设置。其官方训练命令同样采用 `T=4`、`connect_f=ADD`、`lr=0.1` 和 `320 epochs`。

CKDSNN 相关的 distillation loss 以及超参数则按照论文中 Method / Implementation Details 的设置加入即可。

**4. 关于 IF 和代码中的 `LIFSpike`**

您这里的理解是合理的，代码中的 `LIFSpike` 命名确实可能造成一些混淆。

对于论文中的大规模 ImageNet / SEW-ResNet 实验，我们采用的是 **IF neuron**，相关设置主要参考 SEW-ResNet 的官方实现。因此在复现这部分实验时，建议直接参考：

`github.com/fangwei123456/Spike-Element-Wise-ResNet`

对应的 SEW-ResNet 配置，即 `T=4 + ADD + IF neuron`。SEW-ResNet 的公开实现以及后续 SpikingJelly 示例中也明确使用 `MultiStepIFNode` 来构建该配置。

代码中出现的 `LIFSpike` 主要与不同实验代码/实现分支有关，因此不建议仅根据 class name 判断论文中某项实验实际使用的 neuron。对于论文中的 ImageNet 大规模结果，请以论文和 SEW-ResNet reference implementation 中的 **IF setting** 为准。

也许下面两个 repository 会对您的复现有所帮助：

* `github.com/brain-intelligence-lab/temporal_efficient_training`
* `github.com/fangwei123456/Spike-Element-Wise-ResNet`

再次感谢您对我们工作的关注！
